### PR TITLE
[JENKINS-64516] Add support for Robot Framework 4.0

### PR DESCRIPTION
**Pull request by asimell**
Robot Framework 4.0 adds a new test status "SKIP" and removes test criticality. RF 4.0 also removes `<tags></tags>` and `<arguments></arguments>` tags from output.xml, so parser needs to be able to handle new output.xml formats.

This PR
- Adds support for `SKIP` status to job list view, test result overview, Blue Ocean, test result details view
- Adds support for reading both RF 3.x and 4.x styled output.xml files
- Improves some Java syntax related little things
- Updates documentation with fresh screenshots and information on how to configure the plugin for RF 4.0 (use `onlyCritical: false`)
- Improves readability for the list column view by adding a 1 pixel white stroke around the text
- Adds unit tests to test parsing RF 4.x output.xml files.

### CODE REVIEW

File: src/main/java/hudson/plugins/robot/view/RobotListViewColumn.java [hunk 1]
```diff
@@ -41,12 +41,26 @@ public long getTotal(Item job){
...
+
 	public double getPassPercent(Item job) {
 		RobotResult lastRobotResult = getLastRobotResult(job);
 		if (lastRobotResult==null) return 100;
 		return lastRobotResult.getPassPercentage();
 	}
 
+	public double getSkipPercent(Item job) {
+		RobotResult lastRobotResult = getLastRobotResult(job);
+		if (lastRobotResult == null) return 0;
```
Comment (Tattoo) [inline src/main/java/hudson/plugins/robot/view/RobotListViewColumn.java:60]: Please use braces :)

File: src/main/java/hudson/plugins/robot/view/RobotListViewColumn.java [hunk 2]
```diff
@@ -41,12 +41,26 @@ public long getTotal(Item job){
...
+		RobotResult lastRobotResult = getLastRobotResult(job);
+		if (lastRobotResult != null) {
+			return lastRobotResult.getOverallSkipped();
+		}
+		return 0;
+	}
+
 	public double getPassPercent(Item job) {
 		RobotResult lastRobotResult = getLastRobotResult(job);
 		if (lastRobotResult==null) return 100;
```
Comment (Tattoo) [inline src/main/java/hudson/plugins/robot/view/RobotListViewColumn.java:54]: Please use braces :)

**Review by Tattoo — CHANGES_REQUESTED** [review]

File: src/main/java/hudson/plugins/robot/RobotParser.java [hunk 3]
```diff
@@ -357,40 +334,164 @@ private RobotCaseResult processTest(XMLStreamReader reader, RobotSuiteResult res
...
+							stackTrace.append(processForLoop(reader, nestedCount+1));
+							break;
+						case "kw":
+							stackTrace.append(processKeyword(reader, nestedCount+1));
+							break;
+						case "if":
+							stackTrace.append(processIf(reader, nestedCount+1));
+							break;
+						default:
+							break;
```
[the remark this answers is not in the document]
Comment (asimell) [reply src/main/java/hudson/plugins/robot/RobotParser.java:440]: We do get here. There can be `doc`, `msg` and `status` tags.

File: src/main/java/hudson/plugins/robot/model/RobotResult.java [hunk 4]
```diff
@@ -425,6 +445,11 @@ public int getPassed() {
 		return (int) getOverallPassed();
 	}
 
+	@Override
+	public int getSkipped() {
+		return (int)getOverallSkipped();
+	}
+
 	public Api getApi() { return new Api(this); }
```
[the remark this answers is not in the document]
Comment (asimell) [reply src/main/java/hudson/plugins/robot/model/RobotResult.java:454]: That can of course be added. It's just not "necessary". And this plugin uses `getSkipped` as well (in the Jelly files).

File: src/main/java/hudson/plugins/robot/view/RobotListViewColumn.java (hunk 1 above)
Comment (asimell) [reply src/main/java/hudson/plugins/robot/view/RobotListViewColumn.java:60]: How about a few lines up, where this is not used (not changed in this PR).

File: src/main/java/hudson/plugins/robot/RobotParser.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (asimell) [reply src/main/java/hudson/plugins/robot/RobotParser.java:446]: And I don't believe adding empty cases for (at least) `doc`, `msg` and `status` will make it more maintainable. The only fields we're interested in are the ones currently handled, so if the schema gets changed it means more work here if we need to explicitly handle all possible tags.

File: src/main/java/hudson/plugins/robot/model/RobotResult.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Tattoo) [reply src/main/java/hudson/plugins/robot/model/RobotResult.java:454]: So will we do it in this PR or somewhere else?

File: src/main/java/hudson/plugins/robot/view/RobotListViewColumn.java (hunk 1 above)
Comment (Tattoo) [reply src/main/java/hudson/plugins/robot/view/RobotListViewColumn.java]: Please use it there as well :)

Comment (asimell) [reply src/main/java/hudson/plugins/robot/view/RobotListViewColumn.java]: These are outdated. All these are changed already.

File: src/main/java/hudson/plugins/robot/model/RobotResult.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (asimell) [reply src/main/java/hudson/plugins/robot/model/RobotResult.java:454]: I'd say let's do it later.

File: src/main/java/hudson/plugins/robot/RobotParser.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Tattoo) [reply src/main/java/hudson/plugins/robot/RobotParser.java:452]: fair enough

**Review by Tattoo — COMMENTED** [review]
Let's finish the unresolved questions

File: src/main/java/hudson/plugins/robot/view/RobotListViewColumn.java (hunk 1 above)
Comment (Tattoo) [reply src/main/java/hudson/plugins/robot/view/RobotListViewColumn.java]: Right, got confused, my bad. You resolved it the other way I thought

File: src/main/java/hudson/plugins/robot/model/RobotResult.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Tattoo) [reply src/main/java/hudson/plugins/robot/model/RobotResult.java:454]: Check

**Comment by Tattoo** [comment]
LGTM!

**Review by Tattoo — APPROVED** [review]

**Comment by emakaay** [comment]
Nice work, updating the plugin to support robotframework 4.0

I did some initial tests, and found a glitch.

I first tested the plugin(v3.0.0) using un-adjusted robot projects (using robotframework v3.2.2 and robotframework-pabot v1.10.1). That worked perfect (the generated report was processed correctly. Rebot was used creating the report.

Then I tested the plugin(v3.0.0) using un-adjusted robot projects(using robotframework v4.0.0 and robotframework-pabot v1.11). The generated report (created using rebot) was generated without a problem, But the plugin did not pickup the correct values for passed and failed tests

The project had 333 tests in total, 326 passed and 7 failed.

the plugin gave the following result:
                   total,   failed, passed, skipped, pass%
critical tests      0         0            0           0     100%
all tests         337         7        330           0      97.9

in short:
robotframework 3.3.2 without pabot (using rebot) is working
robotframework 4.0.0 without pabot (using rebot) is working
robotframework 3.3.2 with pabot <1.11 (using rebot) is working
robotframework 4.0.0 with pabot =1.11 (using rebot) is not working

I don't know why and i can't share the report for debugging, I'm sorry to say

**Comment by Tattoo** [comment]
@emakaay ,

Tried with the same version combos and smaller test count, and it seems to work correctly. Could you try running your setup with smaller test set to see if the counts work?

Another thing that puzzles me, is why are you using `rebot` together with `pabot`? `pabot` AFAIK already creates you the combined `output.xml`, `log.html` & `report.html`, you shouldn't need to do it separately.

**Comment by emakaay** [comment]
@Tattoo ,

I will try to run a smaller set of tests, to see if that results to different outcome. That will take a couple of days though.

I know pabot already creates the mentioned files. The reason we use rebot, is because we have tests that sometimes fail. Sometimes caused by fragile tests, but most of the time environment issues cause a perfect test to fail. Those failing test will be rerun (with the robot option--rerunfailed) to get a perfect 'green' test result... 

To combine the multiple files we use rebot!

**Comment by Tattoo** [comment]
@emakaay ,

Ok, that makes sense. Just to double-check, I would advise you to also take a look that each different version of `output.xml` is named differently and what version output.xml is actually configured to the plugin -- maybe it picks up the wrong output.xml after all? Diligently naming all output.xmls with `--output` or putting them in separate folders with `--outputdir` let's you be sure of this.

**Comment by emakaay** [comment]
@Tattoo 
You're absolutely right about the output files. Each file should have a unique file name, to prevent problems.
This process is created in a way unique files are created, and only after the rebot phase, the output.xml file created.
This process has been is use for more then a year, and has not changed to test this new plugin. 

I don't think that is causing any problems

**Comment by emakaay** [comment]
I recently said that the new plugin with robotframework v4.0 was working (no pabot). 
But there IS a difference how the report is shown.

Before v3.0 the plugin would always present 2 lines (1 with critical tests, and another with All Tests)

With v3.0 the first line (critical test) is there, but always filled with 0,0,0, 100%

When using the plugin in combination with pabot this is also true

**Comment by Tattoo** [comment]
Yes, this is by design ex [explained in the plugin documentation](https://plugins.jenkins.io/robot/#user-content-robot-framework-4x-compatibility)
